File: packages/typescript/src/catalogs/sepomex/codigos-postales.ts
/**
 * SEPOMEX - Mexican Postal Codes Catalog
 * Official catalog of Mexican postal codes
 */

import { loadCatalogArray } from '../../utils/catalog-loader';
import { normalizeText } from '../../utils/text';
import type { PostalCode } from '../../types';

export class CodigosPostales {
  private static _data: PostalCode[] | null = null;

  /**
   * Load postal codes data (lazy loading with caching)
   */
  private static getData(): PostalCode[] {
    if (!this._data) {
      this._data = loadCatalogArray<PostalCode>('sepomex/codigos_postales.json');
    }
    return this._data;
  }

  /**
   * Get all postal codes
   */
  static getAll(): PostalCode[] {
    return this.getData();
  }

  /**
   * Get postal code information
   */
  static getByCp(cp: string): PostalCode[] {
    return this.getData().filter((postal) => postal.cp === cp || postal.codigo_postal === cp);
  }

  /**
   * Get state name from postal code
   */
  static getEstado(cp: string): string | undefined {
    const postal = this.getData().find((p) => p.cp === cp || p.codigo_postal === cp);
    return postal?.estado;
  }

  /**
   * Get municipality from postal code
   */
  static getMunicipio(cp: string): string | undefined {
    const postal = this.getData().find((p) => p.cp === cp || p.codigo_postal === cp);
    return postal?.municipio;
  }

  /**
   * Search by municipality (accent-insensitive)
   */
  static getByMunicipio(municipio: string): PostalCode[] {
    const searchTerm = normalizeText(municipio);
    return this.getData().filter((postal) => normalizeText(postal.municipio).includes(searchTerm));
  }

  /**
   * Search by state (accent-insensitive)
   */
  static getByEstado(estado: string): PostalCode[] {
    const searchTerm = normalizeText(estado);
    return this.getData().filter((postal) => normalizeText(postal.estado) === searchTerm);
  }

  /**
   * Search by settlement (asentamiento) - accent-insensitive
   *
   * Example: searchByAsentamiento("Aguilas") will find "Las Águilas"
   */
  static searchByAsentamiento(asentamiento: string): PostalCode[] {
    const searchTerm = normalizeText(asentamiento);
    return this.getData().filter((postal) =>
      normalizeText(postal.asentamiento).includes(searchTerm)
    );
  }

  /**
   * Validate postal code exists
   */
  static isValid(cp: string): boolean {
    return this.getData().some((postal) => postal.cp === cp || postal.codigo_postal === cp);
  }
}

File: packages/dart/lib/src/catalogs/sepomex/codigos_postales.dart
/// SEPOMEX Códigos Postales Catalog
///
/// Provides access to 157,000+ Mexican postal codes with complete address data.
/// Note: This is a very large dataset and uses lazy loading.
library;

import 'package:catalogmx/src/catalogs/base_catalog.dart';

/// SEPOMEX Postal Codes Catalog (157,000+ records)
class SepomexCodigosPostales {
  static List<Map<String, dynamic>>? _data;
  static Map<String, List<Map<String, dynamic>>>? _byCP;
  static Map<String, List<Map<String, dynamic>>>? _byState;

  /// Loads the postal codes data (lazy loading due to large size)
  static void _loadData() {
    if (_data != null) return;

    _data = BaseCatalog.loadJsonDataSync('sepomex/codigos_postales.json');

    // Build lookup maps
    _byCP = {};
    _byState = {};

    for (final code in _data!) {
      final cp = code['codigo_postal'] as String? ?? code['cp'] as String?;
      final state = code['estado'] as String? ?? code['d_estado'] as String?;

      if (cp != null) {
        _byCP![cp] ??= [];
        _byCP![cp]!.add(code);
      }

      if (state != null) {
        _byState![state] ??= [];
        _byState![state]!.add(code);
      }
    }
  }

  /// Gets all postal codes (Warning: Very large dataset, may use significant memory)
  static List<Map<String, dynamic>> getAll() {
    _loadData();
    return List.from(_data!);
  }

  /// Gets postal code information by CP (may return multiple colonias)
  static List<Map<String, dynamic>> getByCP(String codigoPostal) {
    _loadData();
    return _byCP![codigoPostal] ?? [];
  }

  /// Gets all postal codes for a state
  static List<Map<String, dynamic>> getByState(String estado) {
    _loadData();
    return _byState![estado.toUpperCase()] ?? [];
  }

  /// Searches postal codes by colonia name
  static List<Map<String, dynamic>> searchByColonia(String query, {String? codigoPostal}) {
    _loadData();
    final normalized = query.toLowerCase().trim();

    List<Map<String, dynamic>> searchList = _data!;
    if (codigoPostal != null) {
      searchList = getByCP(codigoPostal);
    }

    return searchList.where((code) {
      final colonia = (code['colonia'] as String? ?? code['d_asenta'] as String? ?? '').toLowerCase();
      return colonia.contains(normalized);
    }).toList();
  }

  /// Searches postal codes by municipality
  static List<Map<String, dynamic>> searchByMunicipio(String query) {
    _loadData();
    final normalized = query.toLowerCase().trim();

    return _data!.where((code) {
      final mun = (code['municipio'] as String? ?? code['d_mnpio'] as String? ?? '').toLowerCase();
      return mun.contains(normalized);
    }).toList();
  }

  /// Validates if a postal code exists
  static bool isValid(String codigoPostal) {
    _loadData();
    return _byCP!.containsKey(codigoPostal);
  }

  /// Gets total count of postal codes
  static int get count {
    _loadData();
    return _data!.length;
  }

  /// Clears cached data to free memory
  static void clearCache() {
    _data = null;
    _byCP = null;
    _byState = null;
    BaseCatalog.clearCacheFor('sepomex/codigos_postales.json');
  }
}
